### DISCUSSION TITLE
Debugging .net PDK Plugin - extism_plugin_call terminates the process.

### DISCUSSION THREAD

**Discussion by Simonl9l** [extism/dotnet-pdk#77] [Q&A]
If I am using the .Net Host SDK and the .Net PDK, (including host functions) is there a way to debug the plugin?

Both are in the same solution, but breakpoints in the plugin/PDK implementation won't activate.
 
I'm also using Rider but that should not make any difference? 

when I try in invoke a method on the plugin the process terminates with zero error reporting.

```C#
 _plugin = new Plugin(manifest, _hostFunctions, withWasi: true);
            
 var result = _plugin.Call("initialize","{}", _ctx.Token);
```
The explosion happens when the `LibExtism.extism_plugin_call` is invoked.

**Comment by mhmd-azeez** [comment]
Hi, to make sure the host doesn't crash, you can add a try catch:

```
try {
 _plugin = new Plugin(manifest, _hostFunctions, withWasi: true);
            
 var result = _plugin.Call("initialize","{}", _ctx.Token);
} catch (Exception ex) {
// print exception here, or set a break point
}
```

However, you can't debug a wasm plugin using VS. Instead, you could use Console.Writeline (aka printf debugging). Make sure you enable WASI output to see the logs by seeting `EXTISM_ENABLE_WASI_OUTPUT=1` environment variable:

https://github.com/extism/extism/tree/main/runtime#environment-variables

**Reply by Simonl9l** [reply]
I have the try/catch block, but the process terminates before the catch is hit.

**Reply by mhmd-azeez** [reply]
hmmm, that shouldn't happen. can you share a repro?

P.S: also make sure you have installed the runtime nuget package in the host: https://www.nuget.org/packages/Extism.runtime.all

**Reply by Simonl9l** [reply]
Sorry I can't share the repo. I can perhaps look to fire up a standalone one, unless you have one with .Net SDK/PDK with Host Functions I can use as a template. What I have is cranked from scratch using the docs - but is explicit to our proprietary use case.

Here is the host `csproj`

```XML
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
        <TargetFramework>net8.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="Extism.runtime.all" Version="1.2.0" />
        <PackageReference Include="Extism.Sdk" Version="1.2.0" />
    </ItemGroup>
</Project>
```

Here is the plugin `csproj`:
```XML
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <TargetFramework>net8.0</TargetFramework>
        <RuntimeIdentifier>wasi-wasm</RuntimeIdentifier>
        <OutputType>Exe</OutputType>
        <PublishTrimmed>true</PublishTrimmed>
        <!-- Make sure we create a standalone wasm file for our plugin -->
        <WasmSingleFileBundle>true</WasmSingleFileBundle>
        <WasmBuildNative>true</WasmBuildNative>
        <TrimMode>partial</TrimMode>
        <DebuggerSupport>false</DebuggerSupport>
        <EventSourceSupport>false</EventSourceSupport>
        <UseSystemResourceKeys>true</UseSystemResourceKeys>
        <NativeDebugSymbols>false</NativeDebugSymbols>
    </PropertyGroup>
    <ItemGroup>
      <PackageReference Include="Extism.Pdk" Version="1.0.2" />
    </ItemGroup>
</Project>
```

Is this possibly a runtime issue and how would I know? Should I be using some of the other Env Vars too?

Would be great to have all the documentation have code versions for each SDK/PDK - having `rust` everywhere is not helpful if you are a .Net developer - per https://github.com/extism/extism/tree/main/runtime#environment-variables

Every product that uses Extism will have a predominate host language, that we can [wrap](https://github.com/extism/dotnet-pdk/issues/33) and provide bindings packages to make it easy for our customer to develop their own plugins in their language of choice, but does not mean we all know or are comfortable in `rust`?...

**Reply by mhmd-azeez** [reply]
> Is this possibly a runtime issue and how would I know? Should I be using some of the other Env Vars too?

Without a repro it's difficult to say, but it should work fine. We use the .NET SDK for running the .NET PDK & SDK tests:
- https://github.com/extism/dotnet-pdk/blob/main/tests/Extism.Pdk.WasmTests/KitchenSinkTests.cs
- https://github.com/extism/dotnet-sdk/tree/main/test/Extism.Sdk

Both csprojs seem fine to me. Can you share a code snippet on how you're calling the plugins (similar to the one I shared with you). You can simplify it and remove the proprietary parts. If you add a try catch around the plugin code, it should NOT crash your host

**Comment by Simonl9l** [comment]
@mhmd-azeez thanks...

Given you reply above, I cloned the SDK repo, and tried to run the tests.

I note that some of the `cproj` files are still referencing `net7.0`. `SampleHost.csproj` and the `Extism.Odk.MsBuild.Tests.csproj`. My understanding that that `net8.0` was a minimum requirement?

If I try and run the `Extism.Pdk.WasmTests` I get the following log an note the references to some weird paths like `/System/Volumes/Preboot/Cryptexes/OS/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism.dylib` not on my system:
```
System.DllNotFoundException: Unable to load shared library 'extism' or one of its dependencies. In order to help diagnose loading problems, consider s...

System.DllNotFoundException
Unable to load shared library 'extism' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: 
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism.dylib, 0x0001): tried: '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism.dylib' (no such file), '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism.dylib' (no such file)
dlopen(/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/extism.dylib, 0x0001): tried: '/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/extism.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/extism.dylib' (no such file), '/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/extism.dylib' (no such file)
dlopen(extism.dylib, 0x0001): tried: 'extism.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSextism.dylib' (no such file), '/usr/lib/extism.dylib' (no such file, not in dyld cache), 'extism.dylib' (no such file), '/usr/local/lib/extism.dylib' (no such file), '/usr/lib/extism.dylib' (no such file, not in dyld cache)
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libextism.dylib, 0x0001): tried: '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libextism.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libextism.dylib' (no such file), '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libextism.dylib' (no such file)
dlopen(/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/libextism.dylib, 0x0001): tried: '/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/libextism.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/libextism.dylib' (no such file), '/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/libextism.dylib' (no such file)
dlopen(libextism.dylib, 0x0001): tried: 'libextism.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibextism.dylib' (no such file), '/usr/lib/libextism.dylib' (no such file, not in dyld cache), 'libextism.dylib' (no such file), '/usr/local/lib/libextism.dylib' (no such file), '/usr/lib/libextism.dylib' (no such file, not in dyld cache)
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism, 0x0001): tried: '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism' (no such file), '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/extism' (no such file)
dlopen(/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/extism, 0x0001): tried: '/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/extism' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/extism' (no such file), '/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/extism' (no such file)
dlopen(extism, 0x0001): tried: 'extism' (no such file), '/System/Volumes/Preboot/Cryptexes/OSextism' (no such file), '/usr/lib/extism' (no such file, not in dyld cache), 'extism' (no such file), '/usr/local/lib/extism' (no such file), '/usr/lib/extism' (no such file, not in dyld cache)
dlopen(/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libextism, 0x0001): tried: '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libextism' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libextism' (no such file), '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/8.0.0/libextism' (no such file)
dlopen(/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/libextism, 0x0001): tried: '/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/libextism' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/libextism' (no such file), '/Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/bin/Debug/net8.0/libextism' (no such file)
dlopen(libextism, 0x0001): tried: 'libextism' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibextism' (no such file), '/usr/lib/libextism' (no such file, not in dyld cache), 'libextism' (no such file), '/usr/local/lib/libextism' (no such file), '/usr/lib/libextism' (no such file, not in dyld cache)

   at Extism.Sdk.Native.LibExtism.extism_function_new(String name, ExtismValType* inputs, Int64 nInputs, ExtismValType* outputs, Int64 nOutputs, InternalExtismFunction func, IntPtr userData, IntPtr freeUserData)
   at Extism.Sdk.HostFunction..ctor(String functionName, Span`1 inputTypes, Span`1 outputTypes, IntPtr userData, ExtismFunction hostFunction)
   at Extism.Sdk.HostFunction.FromMethod(String functionName, IntPtr userdata, Action`1 callback)
   at Extism.Pdk.WasmTests.KitchenSinkTests.CreatePlugin(String name, Action`1 config, HostFunction[] hostFunctions) in /Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/KitchenSinkTests.cs:line 141
   at Extism.Pdk.WasmTests.KitchenSinkTests.TestConcat() in /Users/<profile>/Projects/dotnet-pdk/tests/Extism.Pdk.WasmTests/KitchenSinkTests.cs:line 26
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

**Comment by mhmd-azeez** [comment]
@Simonl9l i have updated the WasmTests project to reference the extism runtime nuget package:
https://github.com/extism/dotnet-pdk/pull/78

You should be able to run the tests on your machine now

> I note that some of the cproj files are still referencing net7.0. SampleHost.csproj and the Extism.Odk.MsBuild.Tests.csproj. My understanding that that net8.0 was a minimum requirement?

For the PDK yes, but the SDK supports lower versions too. These projects are not using the Extism PDK

**Comment by Simonl9l** [comment]
@mhmd-azeez  thanks again, I pulled the latest `main`

I had to fix this reference to point to Debug (than Release):
```C#
 private string GetWasmPath(string name)
    {
        return Path.Combine(
            Environment.CurrentDirectory,
            $"../../../../../samples/{name}/bin/Debug/net8.0/wasi-wasm/AppBundle/{name}.wasm");
    }
```

However even with that it explodes wih the logging saying ` Process /usr/local/share/dotnet/dotnet:36318 exited with code '0': Not available` wet running the `KitchenSinkTests.TestConfig` tests.

- I have no idea what is "Not Available"?
- I'm assuming this is environmental but not clear what and have followed the [prerequisites](https://github.com/extism/dotnet-pdk#prerequisites) - or it would not even build?
- Any recommendations...?

To confirm I'm running on an Apple Silicon under under MacOS. This is expected to work?

**Reply by mhmd-azeez** [reply]
@Simonl9l can you provide the full list of commands you're running?

This is how we run our tests in CI:
```
make test
```

**Reply by Simonl9l** [reply]
Ok I ran that and it seems to work (`dotnet test` in the test project folder also does to, but for some reason ruin the individual test in/from JetBrains Rider Test runner does not):
```
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:    10, Skipped:     0, Total:    10, Duration: 2 s - Extism.Pdk.WasmTests.dll (net8.0)

Workload updates are available. Run `dotnet workload list` for more information.

```



Im not sure where this leaves me, but do note in you sample `KitchenSinkTests`, in the `CreatePlugin` there is this code:
```C#
foreach (var function in functions)
{
    function.SetNamespace("env");
}
``` 

and do also note that in the SampleLib, that `env` is specific for the `DllImport`:
```C#
[DllImport("env", EntryPoint = "samplelib_import")]
public static extern void samplelib_import();
```

in my own code, per the docs I have this:
```C#
[DllImport("extism", EntryPoint = "End")]
private static extern void End(ulong messageOffset);
```

I don't follow the significance of the `env` for the dllName as this does not seem to be the assembly name for the `SampleLib`; however mu issue is (so far) calling the plugin, not the host function.

I'll try and set up a standalone sample Host/Plugin class, that is more representative of my code...

**Reply by Simonl9l** [reply]
Actually having run the `make test` the tests also work from Rider test runner...??

**Reply by mhmd-azeez** [reply]
> I don't follow the significance of the env for the dllName as this does not seem to be the assembly name for the SampleLib;

By default the host functions are expected to be in the `extism:host/user` (aka `DllImport("extism")]`) module. Setting the module name to `env` is optional. The tests are just making sure that importing from other non-default namespaces works too.

> I'll try and set up a standalone sample Host/Plugin class, that is more representative of my code...

Okay, let us know if you run into any other issues

**Reply by Simonl9l** [reply]
So for arguments sake, should my plugin host itself be an assembly (as sub assembly) that it not part of the executable host but referenced in the project, but arguably would be if AoT'd, do I need to do anything different with this DllHost name, and per the code in the test why is it `env` if I can't find other references to it?

Also would the be a case of the `Call` otherwise failing - as the SDK?PDK can't hook the plugin up correctly - wild guess - or would I get a specific diagnostic error for that than a crash?

When I run the `make test` it seem to run some code that generates other code per `Extism.Pdk.MSBuild` what it is the significance of this project, as it not referenced the documentation AFAICT. Similarly for `Extism.Pdk.MsBuild.Tests`?

As running these seem to make the test work that was not otherwise done just running the tests.

**Reply by mhmd-azeez** [reply]
Hi,

> So for arguments sake, should my plugin host itself be an assembly (as sub assembly) that it not part of the executable host but referenced in the project, but arguably would be if AoT'd, do I need to do anything different with this DllHost name, and per the code in the test why is it env if I can't find other references to it?

I am not sure if I understand this. But in the case of the PDK, the `DllImport` is for a wasm import, not a native DLL import. So it shouldn't matter if you AOT your host or not

> Also would the be a case of the Call otherwise failing - as the SDK?PDK can't hook the plugin up correctly - wild guess - or would I get a specific diagnostic error for that than a crash?

You should get an exception, not a crash

> When I run the make test it seem to run some code that generates other code per Extism.Pdk.MSBuild what it is the significance of this project, as it not referenced the documentation AFAICT. Similarly for Extism.Pdk.MsBuild.Tests?

The PDK uses an MSBuild task to generate some glue code necessary for making exports work. Right now .NET doesn't have an easy way of exprorting functions for wasi. So we had to build our own MSBuild task. This is an implementation detail

**Reply by Simonl9l** [reply]
OK thank again for following up.

> > So for arguments sake, should my plugin host itself be an assembly (as sub assembly) that it not part of the executable host but referenced in the project, but arguably would be if AoT'd, do I need to do anything different with this DllHost name, and per the code in the test why is it env if I can't find other references to it?
> 
> I am not sure if I understand this. But in the case of the PDK, the `DllImport` is for a wasm import, not a native DLL import. So it shouldn't matter if you AOT your host or not
> 

so my use of the DllName of `extism` is correct per:
```C#
[DllImport("extism", EntryPoint = "End")]   private static extern void End(ulong messageOffset);
```
where the corresponds to a hist function defend as follows:
```C#
new HostFunction(
    "End",
    [/* Message */ ExtismValType.PTR,],
    Array.Empty<ExtismValType>(), 
    IntPtr.Zero,
    (plugin, inputs, _) =>
    {
        var message = plugin.ReadString(inputs[0].v.ptr);
        var task = _sink.End(message); // a bit of the impl. - sends the message on.

         Task.WhenAll([task]).ConfigureAwait(true);
    }),
```
This is in a separate referenced assembly, whist not currently AoT'd, will be.
 
> > Also would the be a case of the Call otherwise failing - as the SDK?PDK can't hook the plugin up correctly - wild guess - or would I get a specific diagnostic error for that than a crash?
> 
> You should get an exception, not a crash
> 

Still Crashing...still don't know why!

> > When I run the make test it seem to run some code that generates other code per Extism.Pdk.MSBuild what it is the significance of this project, as it not referenced the documentation AFAICT. Similarly for Extism.Pdk.MsBuild.Tests?
> 
> The PDK uses an MSBuild task to generate some glue code necessary for making exports work. Right now .NET doesn't have an easy way of exprorting functions for wasi. So we had to build our own MSBuild task. This is an implementation detail

To confirm this is all part of what's built into the PDK assembly/package, and nothing we need to do here when building and hosting plugins. Or do we need to do something here - is so where is this documented?

**Reply by mhmd-azeez** [reply]
> so my use of the DllName of extism is correct per:

Yes, this seems correct

>  Task.WhenAll([task]).ConfigureAwait(true);

I'd be careful about using Tasks inside host functions. Since the callback is not async, exceptions are probably not being propagated properly. Maybe this is the source of the crash?

> To confirm this is all part of what's built into the PDK assembly/package, and nothing we need to do here when building and hosting plugins

Exactly, this is just an implementation details and you shouldn't need to worry about it
